server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -541,6 +554,60 @@ public void run() {
         }
     }
 
+    static class TimeChangeChecker {
+
+        private final long thresholdMillis;
+        private final long thresholdNanos;
+        private long absoluteMillis;
+        private long relativeNanos;
+
+        TimeChangeChecker(long thresholdMillis, long absoluteMillis, long relativeNanos) {
+            this.thresholdMillis = thresholdMillis;
+            this.thresholdNanos = TimeValue.timeValueMillis(thresholdMillis).nanos();
+            this.absoluteMillis = absoluteMillis;
+            this.relativeNanos = relativeNanos;
+        }
+
+        void check(long newAbsoluteMillis, long newRelativeNanos) {
+            if (thresholdMillis <= 0) {
+                return;
+            }
+
+            try {
+                final long deltaMillis = newAbsoluteMillis - absoluteMillis;
+                if (deltaMillis > thresholdMillis) {
+                    final TimeValue delta = TimeValue.timeValueMillis(deltaMillis);
+                    logger.warn("timer thread slept for [{}/{}ms] on absolute clock which is above the warn threshold of [{}ms]",
+                            delta,
+                            deltaMillis,
+                            thresholdMillis);
+                } else if (deltaMillis < 0) {
+                    final TimeValue delta = TimeValue.timeValueMillis(-deltaMillis);
+                    logger.warn("absolute clock went backwards by [{}/{}ms] while timer thread was sleeping",
+                            delta,
+                            -deltaMillis);
+                }
+
+                final long deltaNanos = newRelativeNanos - relativeNanos;
+                if (deltaNanos > thresholdNanos) {
+                    final TimeValue delta = TimeValue.timeValueNanos(deltaNanos);
+                    logger.warn("timer thread slept for [{}/{}ns] on relative clock which is above the warn threshold of [{}ms]",
+                            delta,
+                            deltaNanos,
+                            thresholdMillis);
+                } else if (deltaNanos < 0) {
+                    final TimeValue delta = TimeValue.timeValueNanos(-deltaNanos);
+                    logger.warn("relative clock went backwards by [{}/{}ns] while timer thread was sleeping",
```

Comment:
Nit/comment: You could argue that this deserves a `logger.error` and `assert false;` ... I mean this would be a JVM bug?

Reply:
Yeah ok, see aae4e9d. More likely a kernel bug, given that `System::nanoTime` is pretty much just `clock_gettime`. Consider yourself volunteered to investigate if it ever trips tho...